feat(charmhub): add revision support for info command

leyao-daily opened the pull request:
This PR adds support for the `--revision` flag to the `juju info` command when querying charms from CharmHub. The revision flag must be used together with `--channel` to get information about a specific revision of a charm.

Currently there is no way to get information about a specific revision of a charm using `juju info`. This makes it difficult to inspect the details of older charm revisions.

The change allows users to query information about specific charm revisions like:
```bash
juju info postgresql --channel stable --revision 467
```

## Checklist

<!-- If an item is not applicable, use `~strikethrough~`. -->

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [ ] Go unit tests, with comments saying what you're testing
- [ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- [x] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## QA steps
 1. Try getting info without revision:
 `juju info postgresql --channel stable`

 2. Try getting info with revision:
 `juju info postgresql --channel stable --revision 467`

 3. Verify error when using revision without channel:
`Should error with: "ERROR --revision requires --channel to be specified"`

 4. Verify info shows correct revision details when specified.

## Documentation changes

The `juju info` command help text has been updated to document the new `--revision` flag and its requirement to be used with `--channel`.

## Links

<!-- Link to all relevant specification, documentation, bug, issue or JIRA card. -->

**Launchpad bug:** https://bugs.launchpad.net/juju/+bug/2028468

**Jira card:** JUJU-7401

SimonRichardson reviewed (commented):
Why does it matter if this is the case?

> Verify error when using revision without channel:
> Should error with: "ERROR --revision requires --channel to be specified"

Revisions are independent of channels i.e. you can specify any revision at any point in time.

leyao-daily commented:
> Referen

Maybe I didn’t fully understand this part of the requirement correctly. From the bug description, I noticed that this information is used by users for `juju deploy`. While observing the `juju deploy` behavior, I found that the revision needs to be set with --channel.

I re-checked the results of querying the Charmhub API and confirmed that revision is indeed independent.

So, my understanding is that different channels can share the same revision. Our `juju info` command should reflect this result, and for `juju deploy`, it requires a combination of channel + revision to uniquely identify a package. Is this correct?

leyao-daily commented:
@SimonRichardson
Thank you for your comment! I’ve addressed the issues you mentioned. I adjusted the logic for handling revision to correctly reflect the independent relationship and clarified this in the document. Now, `juju info` support both `juju info <charm> --revision <revision>` and `juju info <charm> --revision <revision>

wallyworld reviewed (changes requested):
There's a couple of small drive bys to fix (plus is the cmd doc correct? We added extra doc to the command which is not in the md file). We had a pre-check job but that must only be in main as I can't see it in 3.6. So just run the make target.

leyao-daily commented on cmd/juju/charmhub/info.go:
```diff
@@ -28,6 +28,9 @@ To see channels supported for only a specific base, use the --base flag.
 separated by @. For example, --base ubuntu@22.04.
 
+Use --revision to display information about a specific revision of the charm.
+The --revision flag must be used together with --channel.
```
I will fix the conflict changes.

> Revision is independent of channel. I think if the user species both, we should return an error. I can't see any value in not displaying the whole channel map for a given revision

For this, users might want to ensure a revision is associated with the expected channel before performing further actions. 
1. There are scenarios where a user may want to verify the existence and details of a specific revision within a specific channel before taking further actions.
2. Supporting both individual and combined queries aligns with how other parameters (e.g., --arch) are processed in similar contexts.
3. Allowing combined queries does not introduce ambiguity, as the expected behavior is straightforward:
	•If both revision and channel are specified, display details of the given revision in the given channel
	•If only one is specified, display results accordingly
What do you think about this? 

In the following case, where revision 59 does not exist in the stable channel, allowing combined queries would let the user easily determine this fact in a single operation.
```
juju info landscape-client --revision 59
name: landscape-client
publisher: Landscape Charmers
summary: Charm for installing landscape-client
description: |
  Landscape is a web-based tool for managing Ubuntu systems. This package
  is necessary if you want your machine to be managed in a Landscape
  via Juju.  This package provides the Landscape client and requires a
  Landscape account.
store-url: https://charmhub.io/landscape-client
charm-id: WsixdI1C4ocYA0eZyXALwrQa8pi5TvDD
supports: ubuntu@22.04, ubuntu@23.10, ubuntu@23.04, ubuntu@20.04
tags: monitoring
subordinate: true
relations:
  provides: {}
  requires:
    container: juju-info
channels: |
  latest/stable:     –
  latest/candidate:  –
  latest/beta:       59  2023-06-13  (59)  6MB  amd64, arm64, ppc64el, riscv64, s390x  ubuntu@22.04, ubuntu@22.10, ubuntu@20.04
  latest/edge:       59  2023-06-13  (59)  6MB  amd64, arm64, ppc64el, riscv64, s390x  ubuntu@22.04, ubuntu@22.10, ubuntu@20.04
  legacy/stable:     –
  legacy/candidate:  –
  legacy/beta:       –
  legacy/edge:       –
  
juju info landscape-client --channel stable
name: landscape-client
publisher: Landscape Charmers
summary: Charm for installing landscape-client
description: |
  Landscape is a web-based tool for managing Ubuntu systems. This package
  is necessary if you want your machine to be managed in a Landscape
  via Juju.  This package provides the Landscape client and requires a
  Landscape account.
store-url: https://charmhub.io/landscape-client
charm-id: WsixdI1C4ocYA0eZyXALwrQa8pi5TvDD
supports: ubuntu@22.04, ubuntu@23.10, ubuntu@23.04, ubuntu@20.04
tags: monitoring
subordinate: true
relations:
  provides: {}
  requires:
    container: juju-info
channels: |
  latest/stable:     69  2023-12-11  (69)  6MB  amd64, arm64, ppc64el, s390x  ubuntu@22.04, ubuntu@23.10, ubuntu@23.04, ubuntu@20.04
  latest/candidate:  ↑
  latest/beta:       ↑
  latest/edge:       ↑
```

leyao-daily commented on cmd/juju/charmhub/info.go:
```diff
@@ -28,6 +28,9 @@ To see channels supported for only a specific base, use the --base flag.
 --base can be specified using the OS name and the version of the OS, 
 separated by @. For example, --base ubuntu@22.04.
```
Fixed.

leyao-daily commented on docs/user/reference/juju-cli/list-of-juju-cli-commands/info.md:
```diff
@@ -18,6 +18,7 @@ Displays detailed information about CharmHub charms.
 | `--config` | false | display config for this charm |
 | `--format` | tabular | Specify output format (json&#x7c;tabular&#x7c;yaml) |
 | `-o`, `--output` |  | Specify an output file |
+| `--revision` |  | specify a revision |
```
Fixed all to start with a capital here. (Specify/Display)

leyao-daily commented on cmd/juju/charmhub/convert_test.go:
```diff
@@ -19,6 +19,11 @@ type filterSuite struct {
 var _ = gc.Suite(&filterSuite{})
 
+const (
+	// NoRevision represents an unspecified revision filter
```
Fixed.

wallyworld commented:
Some further input...
It's ok to accept both `--channel` and `--revision` as input, so long as any combinations that cannot be satisfied when used with a download or deploy command return a not found error.
If a charm is found, we should always display the channel map for the relevant track, even if a channel was specified. It allows the user to see a complete picture and if we just show the asked for channel, the output is misleading as you can't tell the difference between missing and filtered out.
If a given revision exists in say stable, and the user says `--channel=candidate --revision=666` we should still display the charm and channel map.
Note it seems `juju download` does not allow both `--channel` and `--revision`
```
$ juju download postgresql --channel 14/edge --revision 545
ERROR --revision cannot be specified together with --arch, --base, --channel or --series
```
But juju deploy does
```
juju deploy postgresql --channel 14/edge --revision 545
Deployed "postgresql" from charm-hub charm "postgresql", revision 545 in channel 14/edge on ubuntu@22.04/stable
```
But deploy "lies" to the user because the charm has actually come from beta channel.

So before we move on with this, I'd like us to get agreement and clarity on the desired behaviour and we need to ensure info, download, and deploy are consistent

leyao-daily commented:
> Some further input... It's ok to accept both `--channel` and `--revision` as input, so long as any combinations that cannot be satisfied when used with a download or deploy command return a not found error. If a charm is found, we should always display the channel map for the relevant track, even if a channel was specified. It allows the user to see a complete picture and if we just show the asked for channel, the output is misleading as you can't tell the difference between missing and filtered out. If a given revision exists in say stable, and the user says `--channel=candidate --revision=666` we should still display the charm and channel map. Note it seems `juju download` does not allow both `--channel` and `--revision`
> 
> ```
> $ juju download postgresql --channel 14/edge --revision 545
> ERROR --revision cannot be specified together with --arch, --base, --channel or --series
> ```
> 
> But juju deploy does
> 
> ```
> juju deploy postgresql --channel 14/edge --revision 545
> Deployed "postgresql" from charm-hub charm "postgresql", revision 545 in channel 14/edge on ubuntu@22.04/stable
> ```
> 
> But deploy "lies" to the user because the charm has actually come from beta channel.
> 
> So before we move on with this, I'd like us to get agreement and clarity on the desired behaviour and we need to ensure info, download, and deploy are consistent

Thank you for the detailed input! I agree with the need to align the behavior across info, download, and deploy. Regarding the behavior of juju deploy, perhaps we can issue a warning while maintaining the current approach, as this method has been in use until now.

Should we standardize how results are displayed across commands and fix the inconsistent behavior in commands through this PR **or** we first establish a standard, and then focus this PR on ensuring that info adheres to the standard. We can handle other requests that don’t conform to the standard in a subsequent PR.create a new PR marked as BLOCKED?

leyao-daily commented:
@wallyworld Following the meeting discussion, I add the `--track` for `juju info`, report error for specify both `--revision` and `--channel` here is the out put of current implementation:

`juju info postgresql --track 16`

> ```
> name: postgresql
> publisher: Canonical
> summary: Charmed PostgreSQL VM operator
> description: |
>   Charm to operate the PostgreSQL database on machines.
> store-url: https://charmhub.io/postgresql
> charm-id: ChgcZB3RhaDOnhkAv9cgRg52LhjBbDt8
> supports: ubuntu@22.04
> tags: databases
> subordinate: false
> relations:
>   provides:
>     cos-agent: cos_agent
>     database: postgresql_client
>     db: pgsql
>     db-admin: pgsql
>     replication-offer: postgresql_async
>   requires:
>     certificates: tls-certificates
>     replication: postgresql_async
>     s3-parameters: s3
>     tracing: tracing
> channels: |
>   16/stable:     –
>   16/candidate:  –
>   16/beta:       –
>   16/edge:       527  2024-11-27  (527)  32MB  arm64  ubuntu@24.04
> ```

`juju info postgresql --revision 545`

> ```
> name: postgresql
> publisher: Canonical
> summary: Charmed PostgreSQL VM operator
> description: |
>   Charm to operate the PostgreSQL database on machines.
> store-url: https://charmhub.io/postgresql
> charm-id: ChgcZB3RhaDOnhkAv9cgRg52LhjBbDt8
> supports: ubuntu@22.04
> tags: databases
> subordinate: false
> relations:
>   provides:
>     cos-agent: cos_agent
>     database: postgresql_client
>     db: pgsql
>     db-admin: pgsql
>     replication-offer: postgresql_async
>   requires:
>     certificates: tls-certificates
>     replication: postgresql_async
>     s3-parameters: s3
>     tracing: tracing
> channels: |
>   14/stable:     –
>   14/candidate:  545  2025-01-19  (545)  32MB  amd64  ubuntu@22.04
>   14/beta:       545  2025-01-19  (545)  32MB  amd64  ubuntu@22.04
>   14/edge:       ↑
> ```

`juju info postgresql --revision 545 --channel stable`
> ` ERROR --revision cannot be specified together with --arch, --base, --channel or --series`

`juju info postgresql --revision 551 --track 14`

> ```
> name: postgresql
> publisher: Canonical
> summary: Charmed PostgreSQL VM operator
> description: |
>   Charm to operate the PostgreSQL database on machines.
> store-url: https://charmhub.io/postgresql
> charm-id: ChgcZB3RhaDOnhkAv9cgRg52LhjBbDt8
> supports: ubuntu@22.04
> tags: databases
> subordinate: false
> relations:
>   provides:
>     cos-agent: cos_agent
>     database: postgresql_client
>     db: pgsql
>     db-admin: pgsql
>     replication-offer: postgresql_async
>   requires:
>     certificates: tls-certificates
>     replication: postgresql_async
>     s3-parameters: s3
>     tracing: tracing
> channels: |
>   14/stable:     –
>   14/candidate:  –
>   14/beta:       –
>   14/edge:       551  2025-01-23  (551)  26MB  amd64  ubuntu@22.04
> ```

wallyworld commented on cmd/juju/charmhub/info.go:
```diff
@@ -83,6 +94,8 @@ func (c *infoCommand) SetFlags(f *gnuflag.FlagSet) {
 	f.StringVar(&c.channel, "channel", "", "specify a channel to use instead of the default release")
 	f.BoolVar(&c.config, "config", false, "display config for this charm")
+	f.IntVar(&c.revision, "revision", -1, "specify a revision number")
+	f.StringVar(&c.track, "track", "", "specify a track to use instead of the default all tracks")
```
`... instead of the default track`

leyao-daily replied:
fixed.

wallyworld commented on cmd/juju/charmhub/info.go:
```diff
@@ -187,7 +212,7 @@ func (c *infoCommand) Run(cmdContext *cmd.Context) error {
 	}
 
-	view, err := convertInfoResponse(info, c.arch, base)
+	view, err := convertInfoResponse(info, c.arch, c.revision, c.track, base)
```
As we do with channel on line 200, we need to do something similar for track. Track is the same as using a channel of `<track>/stable` (but only if revision isn't specified). ie we need to call `WithInfoChannel` to specify where to look for the charm.

leyao-daily replied:
It makes sense, but I try with that, it seems that query the charmhub api using channel=16/stable for postgresql does not give the right [output](https://api.charmhub.io/v2/charms/info/postgresql?channel=16/stable&fields=result.categories.featured%2Cresult.categories.name%2Cresult.contains-charms.name%2Cresult.contains-charms.package-id%2Cresult.contains-charms.store-url%2Cresult.description%2Cresult.license%2Cresult.publisher.display-name%2Cresult.store-url%2Cresult.summary%2Cdefault-release.revision.download.size%2Cdefault-release.revision.config-yaml%2Cdefault-release.revision.metadata-yaml%2Cdefault-release.revision.bundle-yaml%2Cdefault-release.revision.bases.architecture%2Cdefault-release.revision.bases.name%2Cdefault-release.revision.bases.channel%2Cdefault-release.revision.revision%2Cdefault-release.revision.version%2Cdefault-release.channel.name%2Cdefault-release.channel.base.architecture%2Cdefault-release.channel.base.name%2Cdefault-release.channel.base.channel%2Cdefault-release.channel.released-at%2Cdefault-release.channel.risk%2Cdefault-release.channel.track%2Cchannel-map.revision.download.size%2Cchannel-map.revision.created-at%2Cchannel-map.revision.bases.architecture%2Cchannel-map.revision.bases.name%2Cchannel-map.revision.bases.channel%2Cchannel-map.revision.revision%2Cchannel-map.revision.version%2Cchannel-map.channel.name%2Cchannel-map.channel.base.architecture%2Cchannel-map.channel.base.name%2Cchannel-map.channel.base.channel%2Cchannel-map.channel.released-at%2Cchannel-map.channel.risk%2Cchannel-map.channel.track).

leyao-daily replied:
So my current implementation is to query to get the whole information and filter with specific track.

wallyworld commented on cmd/juju/charmhub/info.go:
```diff
@@ -26,7 +26,16 @@ The charm can be specified by name or by path.
+which cannot be used together with --arch, --base, --channel or --series.
+For example: --revision 42
+
+Use --track to display information about a specific track of the charm,
```
We need some extra text about how track is equivalent to channel of `<track>/stable` etc

leyao-daily replied:
Some difference here is that, for example, `juju info postgresql` 

> with --channel stable
```
channels: |
  14/stable:     468  2024-09-11  (468)  28MB  amd64  ubuntu@22.04
  14/candidate:  ↑
  14/beta:       ↑
  14/edge:       ↑
```
> with --track 14
```
channels: |
  14/stable:     468  2024-09-11  (468)  28MB  amd64  ubuntu@22.04
  14/candidate:  545  2025-01-19  (545)  32MB  amd64  ubuntu@22.04
  14/beta:       545  2025-01-19  (545)  32MB  amd64  ubuntu@22.04
  14/edge:       551  2025-01-23  (551)  26MB  amd64  ubuntu@22.04
```
And we also support
> with --track 16
```
channels: |
  16/stable:     –
  16/candidate:  –
  16/beta:       –
  16/edge:       527  2024-11-27  (527)  32MB  arm64  ubuntu@24.04
```

wallyworld reviewed (commented):
I think this is heading in the right direction

wallyworld commented on cmd/juju/charmhub/convert.go:
```diff
@@ -314,6 +314,15 @@ func filterChannels(channelMap []transport.InfoChannelMap, arch string, base cor
+			continue
+		}
+
+		if revision != -1 && cm.Revision.Revision != revision {
```
If we are to always display the complete channel map for every track that should be included, this needs to be reworked right?

leyao-daily replied:
If we want to display the complete channel map for every track, no adjustment is needed here. The filtering rule here only triggers when a specific track is set, which means we are intentionally looking for all channels within that specific track, and this behavior is correct.

When the goal is to display all tracks, the result from `juju info postgresql` ensures that all track information is properly handled through `tracksSeen` and `trackList`.

wallyworld replied:
I tested the command and this is what I got. The 2nd case is wrong as it doesn't show the complete channel map.
I also want to understand why the charm-id comes out the same both times. It may be I'm missing something.

```
$ juju info postgresql --track 14
name: postgresql
publisher: Canonical
summary: Charmed PostgreSQL VM operator
description: |
  Charm to operate the PostgreSQL database on machines.
store-url: https://charmhub.io/postgresql
charm-id: ChgcZB3RhaDOnhkAv9cgRg52LhjBbDt8
supports: ubuntu@22.04
tags: databases
subordinate: false
relations:
  provides:
    cos-agent: cos_agent
    database: postgresql_client
    db: pgsql
    db-admin: pgsql
    replication-offer: postgresql_async
  requires:
    certificates: tls-certificates
    replication: postgresql_async
    s3-parameters: s3
    tracing: tracing
channels: |
  14/stable:     468  2024-09-11  (468)  28MB  amd64  ubuntu@22.04
  14/candidate:  545  2025-01-19  (545)  32MB  amd64  ubuntu@22.04
  14/beta:       545  2025-01-19  (545)  32MB  amd64  ubuntu@22.04
  14/edge:       551  2025-01-23  (551)  26MB  amd64  ubuntu@22.04
```
```
$ juju info postgresql --channel 14/candidate
name: postgresql
publisher: Canonical
summary: Charmed PostgreSQL VM operator
description: |
  Charm to operate the PostgreSQL database on machines.
store-url: https://charmhub.io/postgresql
charm-id: ChgcZB3RhaDOnhkAv9cgRg52LhjBbDt8
supports: ubuntu@22.04
tags: databases
subordinate: false
relations:
  provides:
    cos-agent: cos_agent
    database: postgresql_client
    db: pgsql
    db-admin: pgsql
    replication-offer: postgresql_async
  requires:
    certificates: tls-certificates
    replication: postgresql_async
    s3-parameters: s3
    tracing: tracing
channels: |
  14/stable:     –
  14/candidate:  545  2025-01-19  (545)  32MB  amd64  ubuntu@22.04
  14/beta:       ↑
  14/edge:       ↑
```

leyao-daily replied:
The 2nd case is because the issue related to the `info --channel` behavior that existed in 3.6 isn't addressed in current PR(I initially thought it would be addressed in the next PR alongside deploy/download fixes). 

Channel is implemented by query charmhub api with `<charmhub-url>?channel=<channel>&[field]. The response from API itself lack of other channel information. And my added track is implemented by query charmhub api with`<charmhub-url>&[field]` and then filter by specific track.

For charm-id, I am not sure why it always same even different channel or revision. I checked with full querying of charmhub api, only one id is shown. It seems to be set by https://github.com/juju/juju/blob/a8418dd22acd34784cba4f5acffaea54a07fbf51/charmhub/filter.go#L32

leyao-daily commented:
Now, it show the complete channel map
> `juju info postgresql --channel 14/candidate`
```
name: postgresql
publisher: Canonical
summary: Charmed PostgreSQL VM operator
description: |
  Charm to operate the PostgreSQL database on machines.
store-url: https://charmhub.io/postgresql
charm-id: ChgcZB3RhaDOnhkAv9cgRg52LhjBbDt8
supports: ubuntu@22.04
tags: databases
subordinate: false
relations:
  provides:
    cos-agent: cos_agent
    database: postgresql_client
    db: pgsql
    db-admin: pgsql
    replication-offer: postgresql_async
  requires:
    certificates: tls-certificates
    replication: postgresql_async
    s3-parameters: s3
    tracing: tracing
channels: |
  14/stable:     553  2025-02-04  (553)  26MB  amd64  ubuntu@22.04
  14/candidate:  553  2025-01-28  (553)  26MB  amd64  ubuntu@22.04
  14/beta:       553  2025-01-28  (553)  26MB  amd64  ubuntu@22.04
  14/edge:       563  2025-02-15  (563)  26MB  amd64  ubuntu@22.04
```
which is same with `--track 14`
```
name: postgresql
publisher: Canonical
summary: Charmed PostgreSQL VM operator
description: |
  Charm to operate the PostgreSQL database on machines.
store-url: https://charmhub.io/postgresql
charm-id: ChgcZB3RhaDOnhkAv9cgRg52LhjBbDt8
supports: ubuntu@22.04
tags: databases
subordinate: false
relations:
  provides:
    cos-agent: cos_agent
    database: postgresql_client
    db: pgsql
    db-admin: pgsql
    replication-offer: postgresql_async
  requires:
    certificates: tls-certificates
    replication: postgresql_async
    s3-parameters: s3
    tracing: tracing
channels: |
  14/stable:     553  2025-02-04  (553)  26MB  amd64  ubuntu@22.04
  14/candidate:  553  2025-01-28  (553)  26MB  amd64  ubuntu@22.04
  14/beta:       553  2025-01-28  (553)  26MB  amd64  ubuntu@22.04
  14/edge:       563  2025-02-15  (563)  26MB  amd64  ubuntu@22.04
```

wallyworld commented:
Doing QA again...
The second case below with a revision shows a channel map that should use the "-" for edge if we're going to filter out the rows where the specified revision doesn't exist.

```
juju info postgresql --track 14
channels: |
  14/stable:     553  2025-02-04  (553)  26MB  amd64  ubuntu@22.04
  14/candidate:  553  2025-01-28  (553)  26MB  amd64  ubuntu@22.04
  14/beta:       553  2025-01-28  (553)  26MB  amd64  ubuntu@22.04
  14/edge:       565  2025-02-20  (565)  26MB  amd64  ubuntu@22.04
```
```
juju info postgresql --track 14 --revision 553
channels: |
  14/stable:     553  2025-02-04  (553)  26MB  amd64  ubuntu@22.04
  14/candidate:  553  2025-01-28  (553)  26MB  amd64  ubuntu@22.04
  14/beta:       553  2025-01-28  (553)  26MB  amd64  ubuntu@22.04
  14/edge:       ↑
```
```
juju info postgresql --track 14 --revision 565
channels: |
  14/stable:     –
  14/candidate:  –
  14/beta:       –
  14/edge:       565  2025-02-20  (565)  26MB  amd64  ubuntu@22.04
```

leyao-daily commented:
> Doing QA again... The second case below with a revision shows a channel map that should use the "-" for edge if we're going to filter out the rows where the specified revision doesn't exist.
> 
> ```
> juju info postgresql --track 14
> channels: |
>   14/stable:     553  2025-02-04  (553)  26MB  amd64  ubuntu@22.04
>   14/candidate:  553  2025-01-28  (553)  26MB  amd64  ubuntu@22.04
>   14/beta:       553  2025-01-28  (553)  26MB  amd64  ubuntu@22.04
>   14/edge:       565  2025-02-20  (565)  26MB  amd64  ubuntu@22.04
> ```
> 
> ```
> juju info postgresql --track 14 --revision 553
> channels: |
>   14/stable:     553  2025-02-04  (553)  26MB  amd64  ubuntu@22.04
>   14/candidate:  553  2025-01-28  (553)  26MB  amd64  ubuntu@22.04
>   14/beta:       553  2025-01-28  (553)  26MB  amd64  ubuntu@22.04
>   14/edge:       ↑
> ```
> 
> ```
> juju info postgresql --track 14 --revision 565
> channels: |
>   14/stable:     –
>   14/candidate:  –
>   14/beta:       –
>   14/edge:       565  2025-02-20  (565)  26MB  amd64  ubuntu@22.04
> ```

We can see from [this section of the code](https://github.com/juju/juju/blob/3.6/cmd/juju/charmhub/infowriter.go#L78-L92) that the current behavior of uparrow follows a specific output strategy—sorting by standard risk levels. If a higher risk level appears, all subsequent risk levels lacking a revision will be displayed as uparrow.

Based on your comment for the second case, do we need to modify this output logic?

wallyworld commented on cmd/juju/charmhub/convert_test.go:
```diff
@@ -19,6 +19,15 @@ type filterSuite struct {
 
+const (
+	// NoRevision represents an unspecified revision filter.
+	NoRevision = -1
```
No need to export these.

leyao-daily replied:
Fixed.

wallyworld reviewed (commented):
It's not quite right.
When no revision is specified,we expect this

```
juju info postgresql
...
channels: |
  14/stable:         553                            2025-02-04  (553)  26MB  amd64  ubuntu@22.04
  14/candidate:      553                            2025-01-28  (553)  26MB  amd64  ubuntu@22.04
  14/beta:           553                            2025-01-28  (553)  26MB  amd64  ubuntu@22.04
  14/edge:           567                            2025-02-20  (567)  26MB  amd64  ubuntu@22.04
  latest/stable:     initial-reactive-278-ge3f064a  2023-11-09  (345)  7MB   amd64  ubuntu@16.04, ubuntu@18.04, ubuntu@20.04, ubuntu@22.04
  latest/candidate:  ↑
  latest/beta:       ↑
  latest/edge:       ↑
  16/stable:         –
  16/candidate:      –
  16/beta:           –
  16/edge:           527  2024-11-27  (527)  32MB  arm64  ubuntu@24.04
```
But instead we get
```
channels: |
  14/stable:         553                            2025-02-04  (553)  26MB  amd64  ubuntu@22.04
  14/candidate:      553                            2025-01-28  (553)  26MB  amd64  ubuntu@22.04
  14/beta:           553                            2025-01-28  (553)  26MB  amd64  ubuntu@22.04
  14/edge:           567                            2025-02-20  (567)  26MB  amd64  ubuntu@22.04
  latest/stable:     initial-reactive-278-ge3f064a  2023-11-09  (345)  7MB   amd64  ubuntu@16.04, ubuntu@18.04, ubuntu@20.04, ubuntu@22.04
  latest/candidate:  –
  latest/beta:       –
  latest/edge:       –
  16/stable:         –
  16/candidate:      –
  16/beta:           –
  16/edge:           527  2024-11-27  (527)  32MB  arm64  ubuntu@24.04
```

wallyworld commented on cmd/juju/charmhub/convert_test.go:
```diff
@@ -333,7 +342,7 @@ func (filterSuite) TestFilterChannels(c *gc.C) {
 	for k, v := range tests {
 		c.Logf("Test %d %s", k, v.Name)
-		_, got, err := filterChannels(v.Input, v.Arch, v.Base)
+		_, got, err := filterChannels(v.Input, v.Arch, noRisk, noRevision, noTrack, v.Base)
```
Do we have enough test coverage here? We're using table tests but just passing one a fixed set of values for risk, rev, track etc.

leyao-daily replied:
Fixed.

wallyworld reviewed (approved):
Thank you for the changes. 
What we have is now a nice improvement and we can land to make progress and tweak if needed.
Final thing to check is do we have the necessary test coverage as per the comment inline.

leyao-daily commented:
/merge
